docs/AUTH_PROTOCOL.md

```diff
@@ -0,0 +1,393 @@
+
+Once a flow is started, the authentication "conversation" begins. At a high level, a conversation looks like the following:
+
+1. The authenticator requests a primary authentication mechanism from the frontend
+2. The frontend displays the appropriate UI for this authentication mechanism
+3. The user interacts with the UI to give it some data (types in a password, touches a fingerprint sensor, etc)
+4. Once the user is finished, the frontend sends the user's entered data to the authenticator
+5. If the user has failed to authenticate, or if there are no more steps left in the flow, the authenticator ends the conversation and returns success or failure
+6. Otherwise, the authenticator continues with the next step of the flow. Return to step 1
+
+Note that the use of this protocol encapuslates one authentication attempt. A flow might have multiple steps in it (i.e. "password" followed by "otp"), and the authenticator's ability to switch primary authentication mechansims is designed just for that. The authenticator should not handle a primary authentication method's failure on its own. For example: if the user enters an incorrect password, the authenticator should not prompt for a second attempt. Instead, it should emit the `authenticationFailed` event and terminate the conversation. Conversely, the authenticator _is_ expected to handle secondary authentication method failures on its own, because these are not fatal to the overall login attempt.
+
+Each authentication mechanism is started via an event that the authenticator sends. The following is a description of each such event and what the frontend should do in response:
+
+- `password`: A generic password box
+    - This method is always a primary method
+    - Arguments (data that the authenticator includes with the event it sends to the frontend):
+        - `overridePrompt`: Overrides the default prompt in the password entry box. Optional. Should only be used if absolutely necessary
+    - Response: frontend emits the `response` event with the following arguments:
+        - `password`: The password entered by the user
+
+- `text`: Arbitrary text input
+    - This method is always primary
+    - Arguments:
+        - `prompt`: The prompt to put in the text box. Required
+    - Response: `response` event with arguments:
+        - `text`: The text entered by the user
+
+- `newPassword`: UI for choosing a new password
+    - The frontend prompts for the user to enter a new password twice, and only responds once the two instances of the new password match.
```

Comment:
This prevent the backend from doing password quality checks on the first password and immediately reject, forcing users to type twice only to find the password is not acceptable. It is terrible UI.
Besides the frontend should never make any decision on what the user insert, even making a decision on "equality" maybe bad, unless the backend has a way to communicate that "equality" is what it is looking for. In general frontend should make no decisions at all on any user input.

Reply:
Resolving in favor of the conversation here: https://github.com/systemd/systemd/pull/39855#discussion_r2558091314